pennylane/transforms/optimization/cancel_inverses.py

```diff
@@ -171,3 +175,79 @@ def qfunc(x, y, z):
+    """Quantum function transform to remove any operations that are applied next to their
+    (self-)inverses or adjoint.
+
+    Args:
+        qfunc (function): A quantum function.
+
+    Returns:
+        function: the transformed quantum function
+
+    **Example**
+
+
+    """
+    # Make a working copy of the list to traverse
+    list_copy = tape.operations.copy()
+    operations = []
+    while len(list_copy) > 0:
+        current_gate = list_copy[0]
+        list_copy.pop(0)
+
+        # Find the next gate that acts on at least one of the same wires
+        next_gate_idx = find_next_gate(current_gate.wires, list_copy)
+
+        # If no such gate is found queue the operation and move on
+        if next_gate_idx is None:
+            operations.append(current_gate)
+            continue
+
+        # Otherwise, get the next gate
+        next_gate = list_copy[next_gate_idx]
+
+        # If either of the two flags is true, we can potentially cancel the gates
+        if _are_inverses(current_gate, next_gate):
+            # If the wires are the same, then we can safely remove both
+            if current_gate.wires == next_gate.wires:
+                list_copy.pop(next_gate_idx)
+                continue
+            # If wires are not equal, there are two things that can happen.
+            # 1. There is not full overlap in the wires; we cannot cancel
+            if len(Wires.shared_wires([current_gate.wires, next_gate.wires])) != len(
+                current_gate.wires
+            ):
+                operations.append(current_gate)
+                continue
+
+            # 2. There is full overlap, but the wires are in a different order.
+            # If the wires are in a different order, gates that are "symmetric"
+            # over all wires (e.g., CZ), can be cancelled.
+            if current_gate in symmetric_over_all_wires:
+                list_copy.pop(next_gate_idx)
+                continue
+            # For other gates, as long as the control wires are the same, we can still
+            # cancel (e.g., the Toffoli gate).
+            if current_gate in symmetric_over_control_wires:
+                # TODO[David Wierichs]: This assumes single-qubit targets of controlled gates
+                if (
+                    len(Wires.shared_wires([current_gate.wires[:-1], next_gate.wires[:-1]]))
+                    == len(current_gate.wires) - 1
+                ):
+                    list_copy.pop(next_gate_idx)
+                    continue
+        # Add gate any cases where
+        # - there is no wire symmetry
+        # - the control wire symmetry does not apply because the control wires are not the same
+        # - neither of the flags are_self_inverses and are_inverses are true
+        operations.append(current_gate)
+        continue
+
+    # Queue the measurements normally
+
+    new_tape = QuantumTape(operations, tape.measurements)
+    return [new_tape], lambda x: x[0]
```

Comment:
Leaving these comments now since this will be used as a template for other transforms.

Since this is source code and not user code, we should push for code to be easy to inspect, log, and traceback, at the expense of slightly more lines.

Let's just compare:
```
>>> ? fn
Signature:  fn(res)
Docstring: <no docstring>
File:      /var/folders/k1/0v_kvphn55lgf_45kntf1hqm0000gq/T/ipykernel_14960/826151872.py
Type:      function
```

```
>>> ? fn
Signature:  fn(results: 'ResultBatch') -> 'Result'
Docstring:
An null postprocessing function returned by transform ``cancel_inverses``.

Args:
    results (ResultBatch): the results from the batch of circuits output by the
      ``cancel_inverses`` transform

Returns:
    Result: The Execution result for a single tape inputted to ```cancel_inverses``.
    
Closure:
    This post-processing does not depend on anything from the outer scope.
    
File:      /var/folders/k1/0v_kvphn55lgf_45kntf1hqm0000gq/T/ipykernel_14960/99110884.py
Type:      function
```
 
This will be especially crucial since these functions will be passed around and called in a completely different place than where they were defined.